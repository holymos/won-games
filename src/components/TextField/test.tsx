import { screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { EmailOutline } from "@styled-icons/evaicons-outline/EmailOutline";

import { renderWithTheme } from "utils/tests/helpers";

import { TextField } from ".";

describe("<TextField />", () => {
  it("Renders with Label", () => {
    renderWithTheme(<TextField label="Label" labelFor="Field" id="Field" />);

    expect(screen.getByLabelText("Label")).toBeInTheDocument();
  });

  it("Renders without Label", () => {
    renderWithTheme(<TextField />);

    expect(screen.queryByLabelText("Label")).not.toBeInTheDocument();
  });

  it("Renders with placeholder", () => {
    renderWithTheme(<TextField placeholder="hey you" />);

    expect(screen.getByPlaceholderText("hey you")).toBeInTheDocument();
  });

  it("Changes its value when typing", async () => {
    const onInput = jest.fn();

    renderWithTheme(
      <TextField
        onInput={onInput}
        label="TextField"
        labelFor="TextField"
        id="TextField"
      />
    );

    const input = screen.getByRole("textbox");
    const text = "This is my new text";
    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).toHaveValue(text);
      expect(onInput).toHaveBeenCalledTimes(text.length);
    });

    expect(onInput).toHaveBeenCalledWith(text);
  });

  it("Is accessible by tab", () => {
    renderWithTheme(
      <TextField label="TextField" labelFor="TextField" id="TextField" />
    );

    const input = screen.getByLabelText("TextField");
    expect(document.body).toHaveFocus();

    userEvent.tab();
    expect(input).toHaveFocus();
  });

  it("should render an icon version", () => {
    renderWithTheme(<TextField icon={<EmailOutline data-testid="icon" />} />);

    expect(screen.getByTestId("icon")).toBeInTheDocument();
  });

  it("should render an icon on the right side", () => {
    renderWithTheme(
      <TextField
        icon={<EmailOutline data-testid="icon" />}
        iconPosition="right"
      />
    );

    expect(screen.getByTestId("icon").parentElement).toHaveStyle({ order: 1 });
  });

  it("should not change value when disabled", async () => {
    const onInput = jest.fn();

    renderWithTheme(
      <TextField
        onInput={onInput}
        label="Text Field"
        labelFor="textField"
        id="textField"
        disabled
      />
    );

    const input = screen.getByRole("textbox");
    expect(input).toBeDisabled();

    const text = "This is my new text";
    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).not.toHaveValue(text);
    });

    expect(onInput).not.toHaveBeenCalled();
  });

  it("should not be accessible by tab  when disabled", async () => {
    renderWithTheme(
      <TextField
        label="Text Field"
        labelFor="textField"
        id="textField"
        disabled
      />
    );

    const input = screen.getByRole("textbox");
    expect(document.body).toHaveFocus();

    userEvent.tab();

    expect(input).not.toHaveFocus();
  });

  it("renders with error", () => {
    const { container } = renderWithTheme(
      <TextField
        icon={<EmailOutline />}
        label="Text Field"
        labelFor="textField"
        id="textField"
        error="erro"
      />
    );

    const error = screen.getByText(/erro/i);
    const input = screen.getByRole("textbox");

    expect(error).toBeInTheDocument();
    expect(input.parentElement).toHaveStyle({ borderColor: "#FF6347" });
    expect(container).toMatchSnapshot();
  });
});
