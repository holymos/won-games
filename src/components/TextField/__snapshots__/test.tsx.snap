// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> renders with error 1`] = `
.c7 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid;
  border-color: #EAEAEA;
  margin-top: 0.8rem;
}

.c4:focus-within {
  box-shadow: 0 0 0 0.4rem rgba(242,49,165,0.3);
}

.c8 {
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem 0;
  padding-left: 1.6rem;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c2 {
  font-size: 1.4rem;
  color: #030517;
  cursor: pointer;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 2.2rem;
  color: #8F8F8F;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
}

.c6 > svg {
  width: 100%;
}

.c9 {
  color: #FF6347;
  font-size: 1.2rem;
}

.c0 .c3 {
  border-color: #FF6347;
}

.c0 .c1,
.c0 .c5 {
  color: #FF6347;
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1 c2"
      for="textField"
    >
      Text Field
    </label>
    <div
      class="c3 c4"
    >
      <div
        class="c5 c6"
      >
        <svg
          aria-hidden="true"
          class="c7"
          fill="currentColor"
          focusable="false"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            data-name="Layer 2"
          >
            <path
              d="M19 4H5a3 3 0 00-3 3v10a3 3 0 003 3h14a3 3 0 003-3V7a3 3 0 00-3-3zm-.67 2L12 10.75 5.67 6zM19 18H5a1 1 0 01-1-1V7.25l7.4 5.55a1 1 0 00.6.2 1 1 0 00.6-.2L20 7.25V17a1 1 0 01-1 1z"
              data-name="email"
            />
          </g>
        </svg>
      </div>
      <input
        class="c8"
        id="textField"
        type="text"
        value=""
      />
    </div>
    <p
      class="c9"
    >
      erro
    </p>
  </div>
</div>
`;
